### PR TITLE
Add user metadata cont

### PR DESCRIPTION
**Pull request by ahuds001**
## Describe the big picture - why are we doing this?
To allow users to leverage ``user_metadata`` in the ``HybridModel`` and ``HybridPretrainedModel`` as requested in https://github.com/ShopRunner/collie/issues/21

## How is this tested?
Existing tests have been extended and additional tests have been written to make sure that the models are performing as expected.

## Pull Request Checklist
 - [X] Pull request includes a description of why we are doing this
 - [X] ``CHANGELOG`` has been updated
 - [X] Version in ``_version.py`` has been updated
 - [X] All tests in the ``tests`` folder pass with a local build
 - [ ] ``README`` has been updated (if applicable)
 - [X] Documentation in ``docs`` has been updated (if applicable)
 - [ ] ``requirements.txt`` and ``requirements-dev.txt`` have been recompiled (if applicable)
 - [X] Docker image can be built using ``docker build -t collie .``

### CODE REVIEW

**Comment by ahuds001** [comment]
@nathancooperjones  not sure what to do about the tests failure:

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
Overall, this is fantastic! Such a useful change that I have been hoping to make for so long. Incredible that this is finally done and actual functioning code. 

Looking through the entirety of the changes with fresh eyes, I have a couple of suggestions of how we can simplify this code a bit more. 

In addition, it looks like our test coverage checks have failed here. We might need to add some more test coverage. Ideally, every line of code in Collie will be tested in at least one unit test. We can check those results with: 

```python
pytest --cov-report term-missing --cov=collie
```

Which will show you what lines are missed in the PR. 

Let me know how I can help at all with getting this PR merged in 🎉

**Comment by nathancooperjones** [comment]
For context to future reviews, I already took a first pass at this PR here: https://github.com/ahuds001/collie/pull/2

**Comment by nathancooperjones** [comment]
> Never mind, seems to have fixed itself on rerun.
> 
> ~@nathancooperjones not sure what to do about the tests failure. I didn't touch this test or anything around it as far as I can tell, so not sure what is happening?~
> 
> ```
>     def test_explicit_models_trained_for_one_step(explicit_models_trained_for_one_step,
>                                                   train_val_explicit_data):
>         train, _ = train_val_explicit_data
>     
>         item_preds = explicit_models_trained_for_one_step.get_item_predictions(user_id=0,
>                                                                                unseen_items_only=True,
>                                                                                sort_values=True)
>     
>         assert isinstance(item_preds, pd.Series)
>         assert len(item_preds) > 0
>         assert len(item_preds) < len(train)
>     
>         user_preds = explicit_models_trained_for_one_step.get_user_predictions(item_id=0,
>                                                                                unseen_users_only=True,
>                                                                                sort_values=True)
>     
>         assert isinstance(user_preds, pd.Series)
>         assert len(user_preds) > 0
>         assert len(user_preds) < len(train)
>     
>         item_similarities = explicit_models_trained_for_one_step.item_item_similarity(item_id=42)
>         user_similarities = explicit_models_trained_for_one_step.user_user_similarity(user_id=42)
>     
>         if not isinstance(explicit_models_trained_for_one_step, ColdStartModel):
>             # cold start models aren't trained enough for this check to be true
> >           assert item_similarities.index[0] == 42
> E           assert 77 == 42
> ```

This happens sometimes, and it's super annoying. Usually a re-run will fix the issue. Eventually, I will make some time to investigate and fix this!

File: tutorials/05_hybrid_model.ipynb [hunk 1]
```diff
@@ -21,7 +21,7 @@
   },
```
Comment (nathancooperjones) [inline tutorials/05_hybrid_model.ipynb:21]: <p>[Linting] Alphabetize imports: </p><p> </p><pre class="ql-syntax" spellcheck="false">from collie.movielens import get_movielens_metadata, get_recommendation_visualizations, get_user_metadata
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ShopRunner/collie/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: tutorials/06_multi_stage_models.ipynb [hunk 2]
```diff
@@ -25,15 +25,7 @@
    "metadata": {
```
Comment (nathancooperjones) [inline tutorials/06_multi_stage_models.ipynb:25]: <pre class="ql-syntax" spellcheck="false">Line #13.                                  get_user_metadata,
</pre><p>Ditto here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ShopRunner/collie/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: collie/model/hybrid_matrix_factorization.py [hunk 3]
```diff
@@ -213,36 +238,97 @@ def __init__(self,
...
-            joblib.load(os.path.join(load_model_path, 'metadata.pkl'))
-        )
+        try:
+            self.item_metadata = (
+                joblib.load(os.path.join(load_model_path, 'item_metadata.pkl'))
+            )
+        except FileNotFoundError:
+            if self.hparams.item_metadata_layers_dims is not None:
+                warnings.warn('```item_metadata.pkl`` not found')
+        try:
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py:279]: 👍

File: collie/model/hybrid_matrix_factorization.py [hunk 4]
```diff
@@ -213,36 +238,97 @@ def __init__(self,
...
-        self.item_metadata = (
-            joblib.load(os.path.join(load_model_path, 'metadata.pkl'))
-        )
+        try:
+            self.item_metadata = (
+                joblib.load(os.path.join(load_model_path, 'item_metadata.pkl'))
+            )
+        except FileNotFoundError:
+            if self.hparams.item_metadata_layers_dims is not None:
+                warnings.warn('```item_metadata.pkl`` not found')
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py:278]: 👍

File: collie/model/hybrid_matrix_factorization.py [hunk 5]
```diff
@@ -213,36 +238,97 @@ def __init__(self,
...
+
+            setattr(self, f'{metadata_type}_metadata_layers', full_metadata_layers)
+
+            for i, layer in enumerate(getattr(self, f'{metadata_type}_metadata_layers')):
+                nn.init.xavier_normal_(
+                    getattr(self, f'{metadata_type}_metadata_layers')[i].weight
+                )
+                self.add_module(f'{metadata_type}_metadata_layer_{i}', layer)
+
+            setattr(self, f'{metadata_type}_metadata_layers_dims', full_metadata_layers_dims)
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py:330]: 👍

File: collie/model/hybrid_matrix_factorization.py [hunk 6]
```diff
@@ -264,29 +352,48 @@ def _setup_model(self, **kwargs) -> None:
...
+        # set up item metadata-only layers
+        item_metadata_output_dim = self.hparams.item_metadata_num_cols
+        self.item_metadata_layers = None
+        if self.hparams.item_metadata_layers_dims is not None:
+            self._configure_metadata_layers(
+                metadata_type='item',
+                metadata_layers_dims=self.hparams.item_metadata_layers_dims,
+                num_metadata_cols=self.hparams.item_metadata_num_cols,
+            )
+            item_metadata_output_dim = self.item_metadata_layers_dims[-1]
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py:364]: 👍

File: collie/model/hybrid_matrix_factorization.py [hunk 7]
```diff
@@ -296,6 +403,47 @@ def _setup_model(self, **kwargs) -> None:
...
+            Metadata output
+        """
+        # TODO: remove self.device and let lightning do it
+
+        metadata = getattr(self, f'{metadata_type}_metadata')
+        metadata_layers = getattr(self, f'{metadata_type}_metadata_layers')
+        if metadata_type == 'item':
+            metadata_output = metadata[items, :].to(self.device)
+        else:
+            metadata_output = metadata[users, :].to(self.device)
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py:436]: 👍

File: collie/model/hybrid_matrix_factorization.py [hunk 8]
```diff
@@ -296,6 +403,47 @@ def _setup_model(self, **kwargs) -> None:
             nn.init.xavier_normal_(self.combined_layers[i].weight)
             self.add_module('combined_layer_{}'.format(i), layer)
 
+    def _calculate_metadata_output(
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py:406]: 👍

File: collie/model/hybrid_matrix_factorization.py [hunk 9]
```diff
@@ -349,7 +493,58 @@ def forward(self, users: torch.tensor, items: torch.tensor) -> torch.tensor:
...
+                    F.leaky_relu(
+                        combined_nn_layer(combined_output)
+                    )
+                )
+
+            pred_scores = (
+                self.combined_layers[-1](combined_output)
+                + self.user_biases(users)
+                + self.item_biases(items)
+            )
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py:547]: 👍

File: collie/model/hybrid_pretrained_matrix_factorization.py [hunk 10]
```diff
@@ -114,37 +125,102 @@ def __init__(self,
...
     def _load_model_init_helper(self, load_model_path: str, map_location: str, **kwargs) -> None:
-        self.item_metadata = (
-            joblib.load(os.path.join(load_model_path, 'metadata.pkl'))
-        )
+        try:
+            self.item_metadata = (
+                joblib.load(os.path.join(load_model_path, 'item_metadata.pkl'))
+            )
+        except FileNotFoundError:
+            warnings.warn('`item_metadata.pkl` not found')
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_pretrained_matrix_factorization.py:173]: 👍

File: collie/model/hybrid_pretrained_matrix_factorization.py [hunk 11]
```diff
@@ -114,37 +125,102 @@ def __init__(self,
...
+        try:
+            self.user_metadata = (
+                joblib.load(os.path.join(load_model_path, 'user_metadata.pkl'))
+            )
+        except FileNotFoundError:
+            warnings.warn('`user_metadata.pkl` not found')
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_pretrained_matrix_factorization.py:179]: 👍

File: collie/model/hybrid_pretrained_matrix_factorization.py [hunk 12]
```diff
@@ -155,9 +231,10 @@ def _setup_model(self, **kwargs) -> None:
         if self.hparams.load_model_path is None:
             if not hasattr(self, '_trained_model'):
                 self._trained_model = kwargs.pop('trained_model')
-            if not hasattr(self, 'item_metadata'):
+            if not hasattr(self, 'item_metadata') and 'item_metadata' in kwargs:
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_pretrained_matrix_factorization.py:234]: 👍

File: collie/model/hybrid_pretrained_matrix_factorization.py [hunk 13]
```diff
@@ -242,28 +380,74 @@ def forward(self, users: torch.tensor, items: torch.tensor) -> torch.tensor:
...
-        if str(self.device) != str(self.item_metadata.device):
+
+        if (
+            self.item_metadata is not None
+            and str(self.device) != str(self.item_metadata.device)
+        ) or (
+            self.user_metadata is not None
+            and str(self.device) != str(self.user_metadata.device)
+        ):
             self._move_any_external_data_to_device()
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_pretrained_matrix_factorization.py:391]: 👍

File: CHANGELOG.md [hunk 14]
```diff
@@ -3,6 +3,14 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project uses [Semantic Versioning](http://semver.org/).
 
+# [1.2.3] - 2022-8-13
+### Added
+ - movielens ``get_data.py`` now includes ``get_user_metadata``
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply CHANGELOG.md:8]: 👍

File: CHANGELOG.md [hunk 15]
```diff
@@ -3,6 +3,14 @@ All notable changes to this project will be documented in this file.
...
+ - ``HybridModel`` and ``HybridPretrainedModel`` now take additional optional parameters ``user_metadata`` and ``user_metadata_layers_dims``
+### Changed
+- Added ``item_metadata_layers_dims`` and ``user_metadata_layers_dims`` parameters to ``HybridPretrainedModel`` and ``HybridModel`` and deprecated ``metadata_layers_dims``
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply CHANGELOG.md:11]: 👍

File: CHANGELOG.md [hunk 16]
```diff
@@ -3,6 +3,14 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project uses [Semantic Versioning](http://semver.org/).
 
+# [1.2.3] - 2022-8-13
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply CHANGELOG.md:6]: 👍

File: CHANGELOG.md [hunk 17]
```diff
@@ -3,6 +3,14 @@ All notable changes to this project will be documented in this file.
...
+### Added
+ - movielens ``get_data.py`` now includes ``get_user_metadata``
+ - ``HybridModel`` and ``HybridPretrainedModel`` now take additional optional parameters ``user_metadata`` and ``user_metadata_layers_dims``
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply CHANGELOG.md:9]: 👍

File: CHANGELOG.md [hunk 18]
```diff
@@ -3,6 +3,14 @@ All notable changes to this project will be documented in this file.
...
+### Changed
+- Added ``item_metadata_layers_dims`` and ``user_metadata_layers_dims`` parameters to ``HybridPretrainedModel`` and ``HybridModel`` and deprecated ``metadata_layers_dims``
+- Updated notebooks and examples to include usage of ``user_metadata``
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply CHANGELOG.md:12]: 👍

File: collie/model/hybrid_matrix_factorization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py:330]: @nathancooperjones Unfortunately making this change causes an error that I am not quite sure how to solve with loading and saving the model. For now I have reverted back to keeping it as is. I guess another alternative would be to give it a different name.
```
______________________________________________________________________ test_loading_and_saving_hybrid_pretrained_model _______________________________________________________________________

movielens_metadata_df =       genre_action  genre_adventure  genre_animation  genre_children  genre_comedy  genre_crime  ...  decade_40  decad...        0            0  ...          0          0          0          0          0          1

[1682 rows x 28 columns]
user_metadata_df =      age  gender  occupation_administrator  occupation_artist  occupation_doctor  ...  occupation_salesman  occupation...         0                     0                   1                      0                  0

[943 rows x 23 columns]
train_val_implicit_data = (Interactions object with 43929 interactions between 943 users and 1674 items, returning 10 negative samples per inter...ctions object with 11446 interactions between 943 users and 1674 items, returning 10 negative samples per interaction.)
tmpdir = local('/tmp/pytest-of-root/pytest-0/test_loading_and_saving_hybrid0')

    def test_loading_and_saving_hybrid_pretrained_model(movielens_metadata_df,
                                                        user_metadata_df,
                                                        train_val_implicit_data,
                                                        tmpdir):
        train, val = train_val_implicit_data
        implicit_model = MatrixFactorizationModel(train=train, val=val)
    
        model = HybridPretrainedModel(train=train,
                                      val=val,
                                      item_metadata=movielens_metadata_df,
                                      user_metadata=user_metadata_df,
                                      trained_model=implicit_model,
                                      item_metadata_layers_dims=[16, 8],
                                      user_metadata_layers_dims=[16, 8],
                                      freeze_embeddings=True)
        trainer = CollieTrainer(model=model,
                                logger=False,
                                enable_checkpointing=False,
                                max_epochs=1,
                                gpus=int(str(implicit_model.device).startswith('cuda:0')))
        trainer.fit(model)
    
        item_preds_expected = model.get_item_predictions(user_id=42, unseen_items_only=False)
        user_preds_expected = model.get_user_predictions(item_id=42, unseen_users_only=False)
    
        # set up TemporaryDirectory for writing and reading the file in this test
        temp_dir_name = str(tmpdir)
    
        save_model_path = os.path.join(temp_dir_name, 'test_hybrid_pretrained_model_save')
        model.save_model(save_model_path)
>       loaded_model = HybridPretrainedModel(load_model_path=save_model_path)

tests/test_model.py:1280: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
collie/model/hybrid_pretrained_matrix_factorization.py:156: in __init__
    user_metadata_num_cols=user_metadata_num_cols)
collie/model/base/base_pipeline.py:168: in __init__
    **kwargs)
collie/model/hybrid_pretrained_matrix_factorization.py:181: in _load_model_init_helper
    map_location=map_location)
collie/model/base/base_pipeline.py:256: in _load_model_init_helper
    self.load_state_dict(state_dict=loaded_dict['state_dict'])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = HybridPretrainedModel(
  (embeddings): Sequential(
    (0): ScaledEmbedding(943, 30)
    (1): ScaledEmbedding(1674, 30...r(in_features=64, out_features=32, bias=True)
  (combined_layer_3): Linear(in_features=32, out_features=1, bias=True)
)
state_dict = OrderedDict([('embeddings.0.weight', tensor([[-0.0080, -0.0044, -0.0187,  ..., -0.0043, -0.0011,  0.0027],
        [-0....1675,  0.6009,  0.1303,  0.1042,  0.1783,  0.2621,  0.0298, -0.3574]])), ('combined_layer_3.bias', tensor([0.1362]))])
strict = True

    def load_state_dict(self, state_dict: 'OrderedDict[str, Tensor]',
                        strict: bool = True):
        r"""Copies parameters and buffers from :attr:`state_dict` into
        this module and its descendants. If :attr:`strict` is ``True``, then
        the keys of :attr:`state_dict` must exactly match the keys returned
        by this module's :meth:`~torch.nn.Module.state_dict` function.
    
        Args:
            state_dict (dict): a dict containing parameters and
                persistent buffers.
            strict (bool, optional): whether to strictly enforce that the keys
                in :attr:`state_dict` match the keys returned by this module's
                :meth:`~torch.nn.Module.state_dict` function. Default: ``True``
    
        Returns:
            ``NamedTuple`` with ``missing_keys`` and ``unexpected_keys`` fields:
                * **missing_keys** is a list of str containing the missing keys
                * **unexpected_keys** is a list of str containing the unexpected keys
    
        Note:
            If a parameter or buffer is registered as ``None`` and its corresponding key
            exists in :attr:`state_dict`, :meth:`load_state_dict` will raise a
            ``RuntimeError``.
        """
        missing_keys: List[str] = []
        unexpected_keys: List[str] = []
        error_msgs: List[str] = []
    
        # copy state_dict so _load_from_state_dict can modify it
        metadata = getattr(state_dict, '_metadata', None)
        state_dict = state_dict.copy()
        if metadata is not None:
            # mypy isn't aware that "_metadata" exists in state_dict
            state_dict._metadata = metadata  # type: ignore[attr-defined]
    
        def load(module, prefix=''):
            local_metadata = {} if metadata is None else metadata.get(prefix[:-1], {})
            module._load_from_state_dict(
                state_dict, prefix, local_metadata, True, missing_keys, unexpected_keys, error_msgs)
            for name, child in module._modules.items():
                if child is not None:
                    load(child, prefix + name + '.')
    
        load(self)
        del load
    
        if strict:
            if len(unexpected_keys) > 0:
                error_msgs.insert(
                    0, 'Unexpected key(s) in state_dict: {}. '.format(
                        ', '.join('"{}"'.format(k) for k in unexpected_keys)))
            if len(missing_keys) > 0:
                error_msgs.insert(
                    0, 'Missing key(s) in state_dict: {}. '.format(
                        ', '.join('"{}"'.format(k) for k in missing_keys)))
    
        if len(error_msgs) > 0:
            raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
>                              self.__class__.__name__, "\n\t".join(error_msgs)))
E           RuntimeError: Error(s) in loading state_dict for HybridPretrainedModel:
E           	Missing key(s) in state_dict: "item_metadata_layer_2.weight", "item_metadata_layer_2.bias", "user_metadata_layer_2.weight", "user_metadata_layer_2.bias". 
E           	size mismatch for item_metadata_layer_0.weight: copying a param with shape torch.Size([16, 28]) from checkpoint, the shape in current model is torch.Size([28, 28]).
E           	size mismatch for item_metadata_layer_0.bias: copying a param with shape torch.Size([16]) from checkpoint, the shape in current model is torch.Size([28]).
E           	size mismatch for item_metadata_layer_1.weight: copying a param with shape torch.Size([8, 16]) from checkpoint, the shape in current model is torch.Size([16, 28]).
E           	size mismatch for item_metadata_layer_1.bias: copying a param with shape torch.Size([8]) from checkpoint, the shape in current model is torch.Size([16]).
E           	size mismatch for user_metadata_layer_0.weight: copying a param with shape torch.Size([16, 23]) from checkpoint, the shape in current model is torch.Size([23, 23]).
E           	size mismatch for user_metadata_layer_0.bias: copying a param with shape torch.Size([16]) from checkpoint, the shape in current model is torch.Size([23]).
E           	size mismatch for user_metadata_layer_1.weight: copying a param with shape torch.Size([8, 16]) from checkpoint, the shape in current model is torch.Size([16, 23]).
E           	size mismatch for user_metadata_layer_1.bias: copying a param with shape torch.Size([8]) from checkpoint, the shape in current model is torch.Size([16]).

/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py:1483: RuntimeError
```

File: tests/fixtures/model_fixtures.py [hunk 19]
```diff
@@ -636,7 +636,7 @@ def explicit_models_trained_for_one_step(request,
                             val=val,
                             item_metadata=movielens_metadata_df,
                             embedding_dim=10,
-                            metadata_layers_dims=[16, 12],
+                            item_metadata_layers_dims=[16, 12],
```
[the remark this answers is not in the document]
Comment (ahuds001) [reply tests/fixtures/model_fixtures.py:639]: 👍

File: tutorials/05_hybrid_model.ipynb (hunk 1 above)
Comment (ahuds001) [reply tutorials/05_hybrid_model.ipynb:21]: 👍

File: tutorials/06_multi_stage_models.ipynb (hunk 2 above)
Comment (ahuds001) [reply tutorials/06_multi_stage_models.ipynb:25]: 👍

File: collie/model/hybrid_matrix_factorization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie/model/hybrid_matrix_factorization.py]: Hmmm... have you made sure that all mentions of `* _metadata_layers_dims` are referencing the `hparams`?

File: collie/model/hybrid_matrix_factorization.py [hunk 20]
```diff
@@ -275,7 +275,8 @@ def _load_model_init_helper(self, load_model_path: str, map_location: str, **kwa
             )
         except FileNotFoundError:
             if self.hparams.item_metadata_layers_dims is not None:
-                warnings.warn('```item_metadata.pkl`` not found')
+                warnings.warn('``item_metadata.pkl`` not found')
```
Comment (nathancooperjones) [inline collie/model/hybrid_matrix_factorization.py:278]: I think we should add a test here making sure these warnings are being raised when relevant.

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
This all looks great to me, just looks like the test coverage check is still failing. Looks like Codecov automatically reviewed this PR and noted opportunities we can add tests. Once that's done, I'll approve and we'll merge this in.

File: collie/model/hybrid_matrix_factorization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py]: I did. Looking at the original code it appears that we don't update `self.hparams.metadata_layers_dims` but instead create a separate `metadata_layers_dims` variable. I wonder if the issue comes from the fact that  when we store the parameter in`hparams` we break the `self.add_module` step

```
 if self.hparams.metadata_layers_dims is not None:
            metadata_layers_dims = (
                [self.hparams.item_metadata_num_cols] + self.hparams.metadata_layers_dims
            )
            self.metadata_layers = [
                nn.Linear(metadata_layers_dims[idx - 1], metadata_layers_dims[idx])
                for idx in range(1, len(metadata_layers_dims))
            ]
            for i, layer in enumerate(self.metadata_layers):
                nn.init.xavier_normal_(self.metadata_layers[i].weight)
                self.add_module('metadata_layer_{}'.format(i), layer)

            metadata_output_dim = metadata_layers_dims[-1]
```

**Comment by ahuds001** [comment]
Yup, I need to add a couple of tests but that should be straightforward. I'll take care of it tomorrow/Thursday.

File: collie/model/hybrid_matrix_factorization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahuds001) [reply collie/model/hybrid_matrix_factorization.py]: @nathancooperjones Let me know what you want to do here

**Comment by ahuds001** [comment]
@nathancooperjones increased coverage for tests should be good. The last remaining issue is around `self.{}_metadata_layers_dims` instead of `self.hparams.{}_metadata_layers_dims` . Let me know what you think.

File: collie/model/hybrid_matrix_factorization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie/model/hybrid_matrix_factorization.py]: Looking at the code now, it looks like `self.hparams.item_metadata_layers_dims` is automatically saved when we initialize the model to a value that works for training, saving, and loading the model. I don't think we need to overwrite it at all then! Let's just go ahead and remove the `setattr(self, f'{metadata_type}_metadata_layers_dims', full_metadata_layers_dims)` line entirely! 

The other update we have to make is replacing `self.item_metadata_layers_dims -> self.hparams.item_metadata_layers_dims` and `self.user_metadata_layers_dims -> self.hparams.user_metadata_layers_dims` everywhere else in the code!

**Comment by ahuds001** [comment]
@nathancooperjones done 😄

**Comment by ahuds001** [comment]
@nathancooperjones I reran the 30 simulations for the two models to pass on to the test and got significantly lower bounds for the `mapk_score` because of the high std deviation within the 30 model runs.

**Comment by nathancooperjones** [comment]
> @nathancooperjones I reran the 30 simulations for the two models to pass on to the test and got significantly lower bounds for the `mapk_score` because of the high std deviation within the 30 model runs.

Weird... any idea why the results are so much lower now?

File: tests/test_model.py [hunk 21]
```diff
@@ -984,8 +1038,32 @@ def test_different_item_metadata_types_for_hybrid_pretrained_model(implicit_mode
     trainer_3 = CollieTrainer(model=model_3, logger=False, enable_checkpointing=False, max_epochs=1)
     trainer_3.fit(model_3)
 
+    # ensure that we end up with the same ``user_metadata`` regardless of the input type
+    model_4 = HybridPretrainedModel(train=train,
+                                    val=val,
+                                    user_metadata=user_metadata_df,
+                                    trained_model=implicit_model)
+    trainer_4 = CollieTrainer(model=model_4, logger=False, enable_checkpointing=False, max_epochs=1)
+    trainer_4.fit(model_1)
```
Comment (nathancooperjones) [inline tests/test_model.py:1047]: Last change here, I promise: 

```suggestion
    trainer_4.fit(model_4)
```

Same for the test below this.

**Comment by nathancooperjones** [comment]
> > @nathancooperjones I reran the 30 simulations for the two models to pass on to the test and got significantly lower bounds for the `mapk_score` because of the high std deviation within the 30 model runs.
> 
> Weird... any idea why the results are so much lower now?

Oh, I see that these are new tests that you added for a new model. This makes sense that the MAP@10 here is lower than the traditional matrix factorization model given how little we are training the model here and how simple the dataset is! This is all clear to me!

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
WOOHOOOOOOOO!!!!! THIS IS INCREDIBLE!!!!!!!! 🎉 

Last thing ever for this PR and I swear then I am done requesting changes - can we bump the `CHANGELOG` date to today? As soon as that is done I will approve and merge!

**Review by nathancooperjones — APPROVED** [review]
Amazing work. Thank you so _so_ much @ahuds001!!!!!

**Comment by ahuds001** [comment]
@nathancooperjones happy to help! Thanks for all the great code reviews!!

**Pull request merged**